core/src/test/java/de/jplag/reporting/jsonfactory/DirectoryManagerTest.java

```diff
@@ -0,0 +1,61 @@
+    private static final String SAMPLE_NAME_BC = "basecode";
+    private static final String ROOT_DIRECTORY_BC = "A";
+    private static final String FILE_NAME_BC = "TerrainType.java";
+    private static final File FILE_BC = new File(Path.of(BASE_PATH, SAMPLE_NAME_BC, ROOT_DIRECTORY_BC, FILE_NAME_BC).toString());
+    private static final File SUBMISSION_ROOT_BC = new File(Path.of(BASE_PATH, SAMPLE_NAME_BC, ROOT_DIRECTORY_BC).toString());
+
+    // info of FilesAsSubmissions-sample(in abbreviation FAS)
+    private static final String SAMPLE_NAME_FAS = "FilesAsSubmissions";
+    private static final String ROOT_DIRECTORY_FAS = "Submission1.java";
+    private static final String FILE_NAME_FAS = "Submission1.java";
+    private static final File FILE_FAS = new File(Path.of(BASE_PATH, SAMPLE_NAME_FAS, FILE_NAME_FAS).toString());
+    private static final File SUBMISSION_ROOT_FAS = new File(Path.of(BASE_PATH, SAMPLE_NAME_FAS, ROOT_DIRECTORY_FAS).toString());
+
+    // info of basecode-sameNameOfSubdirectoryAndRootdirectory-sample(in abbreviation BC-SOSAR)
+    private static final String SAMPLE_NAME_BC_SOSAR = "basecode-sameNameOfSubdirectoryAndRootdirectory";
+    private static final String ROOT_DIRECTORY_BC_SOSAR = "A";
+    private static final String SUB_DIRECTORY_BC_SOSAR = Path.of("B", "A").toString();
+    private static final String FILE_NAME_BC_SOSAR = "TerrainType.java";
+    private static final File FILE_BC_SOSAR = new File(
+            Path.of(BASE_PATH, SAMPLE_NAME_BC_SOSAR, ROOT_DIRECTORY_BC_SOSAR, SUB_DIRECTORY_BC_SOSAR, FILE_NAME_BC_SOSAR).toString());
+    private static final File SUBMISSION_ROOT_BC_SOSAR = new File(Path.of(BASE_PATH, SAMPLE_NAME_BC_SOSAR, ROOT_DIRECTORY_BC_SOSAR).toString());
+
+    @Test
+    void testCreateDirectoryBC() throws IOException {
+        File expectedFile = new File(Path.of(OUTPUT_PATH, ROOT_DIRECTORY_BC, FILE_NAME_BC).toString());
+        test(expectedFile, OUTPUT_PATH, ROOT_DIRECTORY_BC, FILE_BC, SUBMISSION_ROOT_BC);
+    }
+
+    @Test
+    void testCreateDirectoryFAS() throws IOException {
+        File expectedFile = new File(Path.of(OUTPUT_PATH, ROOT_DIRECTORY_FAS, FILE_NAME_FAS).toString());
+        test(expectedFile, OUTPUT_PATH, ROOT_DIRECTORY_FAS, FILE_FAS, SUBMISSION_ROOT_FAS);
+    }
+
+    @Test
+    void testCreateDirectoryBC_SOSAR() throws IOException {
+        File expectedFile = new File(Path.of(OUTPUT_PATH, ROOT_DIRECTORY_BC_SOSAR, SUB_DIRECTORY_BC_SOSAR, FILE_NAME_BC_SOSAR).toString());
+        test(expectedFile, OUTPUT_PATH, ROOT_DIRECTORY_BC_SOSAR, FILE_BC_SOSAR, SUBMISSION_ROOT_BC_SOSAR);
+    }
+
+    void test(File expectedFile, String output, String rootDirectory, File file, File submissionRoot) throws IOException {
+        File directory = DirectoryManager.createDirectory(output, rootDirectory, file, submissionRoot);
+        Assertions.assertNotNull(directory);
+        Assertions.assertEquals(expectedFile.getPath(), directory.getPath());
+    }
```

Comment:
There are still several issues in there:

- Duplications
- Abbreviations
- Lack of documentation
- Bad naming
- Ignoring the thrown exception
- Composite constants

Thus, overhauled your test case to show you what the expectations for a good test case are. Please review my suggestions below. Try to understand why I implemented it that way and why it is cleaner and more readable. If you have any questions, feel free to ask. Then, apply the suggestion to this PR.

```suggestion
import static org.junit.jupiter.api.Assertions.fail;

import java.io.File;
import java.io.IOException;
import java.nio.file.Path;

import org.junit.jupiter.api.Assertions;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;

import de.jplag.TestBase;

/**
 * Test for the directory manager that persists the results for the report viewer.
 */
class DirectoryManagerTest extends TestBase {
    private static final Path OUTPUT_PATH = Path.of(BASE_PATH, "output", "submissions");

    private static final String SUBMISSION_1 = "A";
    private static final String FILE_PATH_1 = "TerrainType.java";
    private static final String ROOT_1 = "basecode";

    private static final String SUBMISSION_2 = "Submission1.java";
    private static final String ROOT_2 = "FilesAsSubmissions";

    private static final String SUBMISSION_3 = "A";
    private static final Path FILE_PATH_3 = Path.of("B", "A", "TerrainType.java");
    private static final String ROOT_3 = "basecode-sameNameOfSubdirectoryAndRootdirectory";

    @Test
    @DisplayName("test normal submission with file in folder")
    void testCreateDirectoryBasecode() throws IOException {
        testDirectoryManager(ROOT_1, SUBMISSION_1, FILE_PATH_1);
    }

    @Test
    @DisplayName("test single file as submission")
    void testCreateDirectoryFileAsSubmission() throws IOException {
        testDirectoryManager(ROOT_2, SUBMISSION_2, "");
    }

    @Test
    @DisplayName("test same name of subdirectory and root directory")
    void testCreateDirectorySharedName() throws IOException {
        testDirectoryManager(ROOT_3, SUBMISSION_3, FILE_PATH_3.toString());
    }

    /**
     * Test the directory manager for a given scenario.
     * @param rootName is the name of the root folder.
     * @param submissionName is the name of the submission.
     * @param filePath is the path to the file relative to the submission. Empty for single file submissions.
     */
    private static void testDirectoryManager(String rootName, String submissionName, String filePath) {
        File submissionPath = Path.of(BASE_PATH, rootName, submissionName).toFile();
        File fullFilePath = new File(submissionPath, filePath);
        File expectation = new File(OUTPUT_PATH.toFile(), Path.of(submissionName, filePath.isEmpty() ? submissionName : filePath).toString());
        try {
            File directory = DirectoryManager.createDirectory(OUTPUT_PATH.toString(), submissionName, fullFilePath, submissionPath);
            Assertions.assertNotNull(directory);
            Assertions.assertEquals(expectation.getPath(), directory.getPath());
        } catch (IOException e) {
            fail("Directory manager threw an exception:", e);
        }
    }
```

Reply:
Thanks for your sample and suggestions. I will try to understand what you have changed and learn from it.